packages/core/useFetch/index.ts

```diff
@@ -0,0 +1,161 @@
+
+  /**
+   * The raw response of the fetch response
+   */
+  response: Ref<Response | null>
+
+  /**
+   * Any fetch errors that may have occured
+   */
+  error: Ref<any>
+
+  /**
+   * The fetch response body, may either be JSON or text
+   */
+  data: Ref<object | string | null>
+
+  /**
+   * Indicates if the request is currently being fetched.
+   */
+  isFetching: Ref<boolean>
+
+  /**
+   * Indicates if the fetch request is able to be aborted
+   */
+  canAbort: ComputedRef<boolean>
+
+  /**
+   * Abort the fetch request
+   */
+  abort: Fn
+
+  /**
+   * Manually call the fetch
+   */
+  execute: Fn
+}
+
+interface UseFetchOptions {
+  /**
+   * Will automatically run fetch when `useFetch` is used
+   *
+   * @default true
+   */
+  immediate?: boolean
+
+  /**
+   * Will automatically refetch when the URL is changed if the url is a ref
+   *
+   * @default false
+   */
+  refetch?: MaybeRef<boolean>
+}
+
+export function useFetch(url: MaybeRef<string>): UseFetchReturn
+export function useFetch(url: MaybeRef<string>, useFetchOptions: UseFetchOptions): UseFetchReturn
+export function useFetch(url: MaybeRef<string>, options: RequestInit, useFetchOptions?: UseFetchOptions): UseFetchReturn
+
+export function useFetch(url: MaybeRef<string>, ...args: any[]): UseFetchReturn {
+  const supportsAbort = typeof AbortController === 'function'
+
+  let fetchOptions: RequestInit = {}
+  let options: UseFetchOptions = { immediate: true, refetch: false }
+
+  if (args.length > 0) {
+    if ('immediate' in args[0] || 'refetch' in args[0])
+      options = { ...options, ...args[0] }
+    else
+      fetchOptions = args[0]
+  }
+
+  if (args.length > 1) {
+    if ('immediate' in args[1] || 'refetch' in args[1])
+      options = { ...options, ...args[1] }
+  }
+
+  const isFinished = ref(false)
+  const isFetching = ref(false)
+  const status = ref<number | null>(null)
+  const response = ref<Response | null>(null)
+  const error = ref<any>(null)
+  const data = ref<string | object | null>(null)
+
+  const canAbort = computed(() => {
+    return supportsAbort && isFetching.value
+  })
+
+  let controller: AbortController | null
+
+  const abort = () => {
+    if (supportsAbort && controller instanceof AbortController)
+      controller.abort()
+  }
+
+  const execute = () => {
+    isFetching.value = true
+    isFinished.value = false
+    error.value = null
+
+    if (supportsAbort) {
+      controller = new AbortController()
+      fetchOptions = { ...fetchOptions, signal: controller.signal }
+    }
+
+    fetch(unref(url), fetchOptions)
+      .then((fetchResponse) => {
+        response.value = fetchResponse
+        status.value = fetchResponse.status
+
+        const contentType = fetchResponse.headers.get('content-type')
+
+        if (contentType && contentType.includes('application/json'))
+          fetchResponse.json().then(json => data.value = json)
+        else
+          fetchResponse.text().then(text => data.value = text)
```

Comment:
Could we make this usage simpler with some API?

Reply:
Hmm, is there something out there that would allow us to simplify it? I'm not aware of anything, but if you know I'll update it

Reply:
Not really. But maybe keep your implementation and make a `useFetchJson` wrapper with preconfigured header? Moreover, something like `usePost` and `usePostJson`?

Another approach might be similar to [`ky`](https://github.com/sindresorhus/ky)

```ts
useFetch(URL, {}).json()
```

WDYT?

Reply:
Hmm doing more research into the fetch API they actually have more than just `text()` and `json()`, https://developer.mozilla.org/en-US/docs/Web/API/Body#methods

If we were to make a wrapper for all of these we would have `useFetch`, `useFechJson`, `useFetchBlob`, `useFetchFormData`, and `useFetchArrayBuffer`. But I feel like most of these would not be used very often and it would create quite a few methods. Perhaps another option would be to have an option in the config to set how the data is parsed. Something like this perhaps?

```ts
useFetch(url, {
  // Not sure what a good name for this would be
  // It could be defaulted toauto which would try to automatically detect the type from the response headers
  readAs: 'auto' //could also be json, formData, blob, arrayBuffer, or text
})
```

I do like the idea of wrappers for `usePost` and `usePostJson`. If we do that would we also want ones for form-data and multipart form data?